<launch>

  <arg name="use_left_camera" default="false" />
  <arg name="use_right_camera" default="true" />
  <arg name="use_center_camera" default="false" />

  <!-- left camera params -->
  <arg name="camera_left_name" default="camera_left_6mm" />
  <arg name="camera_left_serial" default="123456789" />
  <arg name="camera_left_calibrated" default="1" />
  <arg name="camera_left_calibration" default="file://$(find pointgrey_camera_driver)/calibration/calibration_6mm.yaml" />
  <arg name="camera_left_conf" default="$(find pointgrey_camera_driver)/conf/conf_6mm.yaml" />

  <!-- right camera params -->
  <arg name="camera_right_name" default="camera_right_6mm" />
  <arg name="camera_right_serial" default="123456788" />
  <arg name="camera_right_calibrated" default="1" />
  <arg name="camera_right_calibration" default="file://$(find pointgrey_camera_driver)/calibration/calibration_6mm.yaml" />
  <arg name="camera_right_conf" default="$(find pointgrey_camera_driver)/conf/conf_6mm.yaml" />

  <!-- center camera params -->
  <arg name="camera_center_name" default="camera_center_6mm" />
  <arg name="camera_center_serial" default="123456787" />
  <arg name="camera_center_calibrated" default="1" />
  <arg name="camera_center_calibration" default="file://$(find pointgrey_camera_driver)/calibration/calibration_12mm.yaml" />
  <arg name="camera_center_conf" default="$(find pointgrey_camera_driver)/conf/conf_12mm.yaml" />

  <group ns="$(arg camera_left_name)" if="$(arg use_left_camera)">
    <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" output="screen"/>

    <node pkg="nodelet" type="nodelet" name="camera_nodelet"
          args="load pointgrey_camera_driver/PointGreyCameraNodelet camera_nodelet_manager" >
      <param name="frame_id" value="$(arg camera_left_name)" />
      <param name="serial" value="$(arg camera_left_serial)" />
      <param name="frame_rate" value="12" />
      <param name="camera_info_url" if="$(arg camera_left_calibrated)"
             value="$(arg camera_left_calibration)" />
    </node>

    <node pkg="image_proc" type="image_proc" name="image_proc_node" >
      <!-- from http://wiki.ros.org/image_proc
      ~queue_size (int, default: 5)
        Size of message queue for synchronizing image and camera_info topics. 
        You may need to raise this if images take significantly longer to travel 
        over the network than camera info.  -->
      <param name="queue_size" value="10"/> 
    </node> 

   <node pkg="nodelet" type="nodelet" name="image_proc_crop"
         args="standalone image_proc/crop_decimate" output="screen">
      <rosparam>
        x_offset: 300
        y_offset: 240
        width: 1320
        height: 720
      </rosparam>
      <!-- input topics -->
      <remap from="camera/image_raw" to="/$(arg camera_left_name)/image_rect_color"/>          
      <remap from="camera/image_info" to="/$(arg camera_left_name)/camera_info"/>
      <!-- output topics -->
      <remap from="camera_out/image_raw" to="crop/image_color"/>
      <remap from="camera_out/image_info" to="crop/camera_info"/> 
    </node>

    <node pkg="dynamic_reconfigure" type="dynparam" name="flir_dynparam_node" args="load camera_nodelet $(arg camera_left_conf)"/>

  </group>

  <group ns="$(arg camera_right_name)" if="$(arg use_right_camera)">
    <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" output="screen"/>

    <node pkg="nodelet" type="nodelet" name="camera_nodelet"
          args="load pointgrey_camera_driver/PointGreyCameraNodelet camera_nodelet_manager" >
      <param name="frame_id" value="$(arg camera_right_name)" />
      <param name="serial" value="$(arg camera_right_serial)" />
      <param name="frame_rate" value="12" />
      <param name="camera_info_url" if="$(arg camera_right_calibrated)"
             value="$(arg camera_right_calibration)" />
    </node>

    <node pkg="image_proc" type="image_proc" name="image_proc_node" >
      <param name="queue_size" value="10"/> 
    </node> 

   <node pkg="nodelet" type="nodelet" name="image_proc_crop"
         args="standalone image_proc/crop_decimate" output="screen">
      <rosparam>
        x_offset: 300
        y_offset: 240
        width: 1320
        height: 720
      </rosparam>
      <!-- input topics -->
      <remap from="camera/image_raw" to="/$(arg camera_right_name)/image_rect_color"/>          
      <remap from="camera/image_info" to="/$(arg camera_right_name)/camera_info"/>
      <!-- output topics -->
      <remap from="camera_out/image_raw" to="crop/image_color"/>
      <remap from="camera_out/image_info" to="crop/camera_info"/> 
    </node>

    <node pkg="dynamic_reconfigure" type="dynparam" name="flir_dynparam_node" args="load camera_nodelet $(arg camera_right_conf)"/>

  </group>

  <group ns="$(arg camera_center_name)" if="$(arg use_center_camera)">
    <node pkg="nodelet" type="nodelet" name="camera_nodelet_manager" args="manager" output="screen"/>

    <node pkg="nodelet" type="nodelet" name="camera_nodelet"
          args="load pointgrey_camera_driver/PointGreyCameraNodelet camera_nodelet_manager" >
      <param name="frame_id" value="$(arg camera_center_name)" />
      <param name="serial" value="$(arg camera_center_serial)" />
      <param name="frame_rate" value="12" />
      <param name="camera_info_url" if="$(arg camera_center_calibrated)"
             value="$(arg camera_center_calibration)" />
    </node>

    <node pkg="image_proc" type="image_proc" name="image_proc_node" >
      <param name="queue_size" value="10"/> 
    </node> 

   <node pkg="nodelet" type="nodelet" name="image_proc_crop"
         args="standalone image_proc/crop_decimate" output="screen">
      <rosparam>
        x_offset: 300
        y_offset: 240
        width: 1320
        height: 720
      </rosparam>
      <!-- input topics -->
      <remap from="camera/image_raw" to="/$(arg camera_center_name)/image_rect_color"/>          
      <remap from="camera/image_info" to="/$(arg camera_center_name)/camera_info"/>
      <!-- output topics -->
      <remap from="camera_out/image_raw" to="crop/image_color"/>
      <remap from="camera_out/image_info" to="crop/camera_info"/> 
    </node>

    <node pkg="dynamic_reconfigure" type="dynparam" name="flir_dynparam_node" args="load camera_nodelet $(arg camera_center_conf)"/>

  </group>

</launch>


